File: src/simplace/SQ_Energy_Balance/net/simplace/sim/components/SQ_Energy_Balance/Conductance.java
package net.simplace.sim.components.SQ_Energy_Balance;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class Conductance extends FWSimComponent
{
    private FWSimVariable<Double> vonKarman;
    private FWSimVariable<Double> heightWeatherMeasurements;
    private FWSimVariable<Double> zm;
    private FWSimVariable<Double> zh;
    private FWSimVariable<Double> d;
    private FWSimVariable<Double> plantHeight;
    private FWSimVariable<Double> wind;
    private FWSimVariable<Double> conductance;

    public Conductance(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Conductance(){
        super();
    }
    @Override
    protected void process()
    {
        double t_vonKarman = vonKarman.getValue();
        double t_heightWeatherMeasurements = heightWeatherMeasurements.getValue();
        double t_zm = zm.getValue();
        double t_zh = zh.getValue();
        double t_d = d.getValue();
        double t_plantHeight = plantHeight.getValue();
        double t_wind = wind.getValue();
        double t_conductance;
        double h;
        h = Math.max(10.0d, t_plantHeight) / 100.0d;
        t_conductance = t_wind * Math.pow(t_vonKarman, 2) / (Math.log((t_heightWeatherMeasurements - (t_d * h)) / (t_zm * h)) * Math.log((t_heightWeatherMeasurements - (t_d * h)) / (t_zh * h)));
        conductance.setValue(t_conductance, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        if(aParamIndex == 0 && aDefineOrCheck == TEST_STATE.DEFINE) //before process
        {
            FWSimVariable.setValue(0.67, iFieldMap.get("Conductance.d"), this);
            FWSimVariable.setValue(0.013, iFieldMap.get("Conductance.zh"), this);
            FWSimVariable.setValue(0.13, iFieldMap.get("Conductance.zm"), this);
            FWSimVariable.setValue(124000, iFieldMap.get("Conductance.wind"), this);
            FWSimVariable.setValue(0, iFieldMap.get("Conductance.plantHeight"), this);
        }
        else if(aParamIndex ==  0 && aDefineOrCheck == TEST_STATE.CHECK) //after process
        {
            FWSimVariable.setValue(598.685, iFieldMap.get("Conductance.conductance"), this);
        }
        else return null;
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Conductance(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("vonKarman", "von Karman constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.42, this));
        addVariable(FWSimVariable.createSimVariable("heightWeatherMeasurements", "reference height of wind and humidity measurements", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 10, 2, this));
        addVariable(FWSimVariable.createSimVariable("zm", "roughness length governing momentum transfer, FAO", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.13, this));
        addVariable(FWSimVariable.createSimVariable("zh", "roughness length governing transfer of heat and vapour, FAO", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.013, this));
        addVariable(FWSimVariable.createSimVariable("d", "corresponding to 2/3. This is multiplied to the crop heigth for calculating the zero plane displacement height, FAO", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0, 1, 0.67, this));
        addVariable(FWSimVariable.createSimVariable("plantHeight", "plant Height", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"", 0, 1000, 0, this));
        addVariable(FWSimVariable.createSimVariable("wind", "wind", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"", 0, 1000000, 124000, this));
        addVariable(FWSimVariable.createSimVariable("conductance", "the boundary layer conductance", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"", 0, 10000, null, this));

        return iFieldMap;
    }
}

File: original/Strategies/Conductance.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml;    
using SiriusQualityEnergyBalance.DomainClass;
namespace SiriusQualityEnergyBalance.Strategies
{
    public class Conductance
    {
        public void Estimate(SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s1,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceRate r,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceAuxiliary a,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceExogenous ex)
        {
            try
            {
                CalculateModel(s, s1, r, a, ex);
            }
            catch (Exception exception)
            {
                string msg = "Error in component SiriusQualityEnergyBalance, strategy: " + this.GetType().Name + ": Unhandled exception running model. "+exception.GetType().FullName+" - "+exception.Message;
                throw new Exception(msg, exception);
            }
        }

        private double _vonKarman;
        public double vonKarman
        {
            get { return this._vonKarman; }
            set { this._vonKarman= value; } 
        }
        private double _heightWeatherMeasurements;
        public double heightWeatherMeasurements
        {
            get { return this._heightWeatherMeasurements; }
            set { this._heightWeatherMeasurements= value; } 
        }
        private double _zm;
        public double zm
        {
            get { return this._zm; }
            set { this._zm= value; } 
        }
        private double _zh;
        public double zh
        {
            get { return this._zh; }
            set { this._zh= value; } 
        }
        private double _d;
        public double d
        {
            get { return this._d; }
            set { this._d= value; } 
        }
        public Conductance() { }
    
        private void CalculateModel(SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s, SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s1, SiriusQualityEnergyBalance.DomainClass.EnergyBalanceRate r, SiriusQualityEnergyBalance.DomainClass.EnergyBalanceAuxiliary a, SiriusQualityEnergyBalance.DomainClass.EnergyBalanceExogenous ex)
        {
                //- Name: Conductance -Version: 1.0, -Time step: 1
                //- Description:
    //            * Title: Conductance Model
    //            * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
    //            * Reference:  https://doi.org/10.1016/0168-1923(94)02214-5
    //        
    //            * Institution: New Zealand Institute for Crop and Food Research Ltd.,
    //            New Zealand Institute for Crop and Food Research Ltd.,
    //            New Zealand Institute for Crop and Food Research Ltd.,
    //            New Zealand Institute for Crop and Food Research Ltd.
    //        
    //            * ExtendedDescription: The boundary layer conductance is expressed as the wind speed profile above the
    //            canopy and the canopy structure. The approach does not take into account buoyancy
    //            effects. 
    //        
    //            * ShortDescription: The boundary layer conductance is expressed as the wind speed profile above the
    //            canopy and the canopy structure. The approach does not take into account buoyancy
    //            effects.
    //        
                //- inputs:
    //            * name: vonKarman
    //                          ** description : von Karman constant
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 1
    //                          ** default : 0.42
    //                          ** unit : dimensionless
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //            * name: heightWeatherMeasurements
    //                          ** description : reference height of wind and humidity measurements
    //                          ** parametercategory : soil
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10
    //                          ** default : 2
    //                          ** unit : m
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : parameter
    //            * name: zm
    //                          ** description : roughness length governing momentum transfer, FAO
    //                          ** parametercategory : constant
    //                          ** inputtype : parameter
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 1
    //                          ** default : 0.13
    //                          ** unit : m
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //            * name: zh
    //                          ** description : roughness length governing transfer of heat and vapour, FAO
    //                          ** parametercategory : constant
    //                          ** inputtype : parameter
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 1
    //                          ** default : 0.013
    //                          ** unit : m
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //            * name: d
    //                          ** description : corresponding to 2/3. This is multiplied to the crop heigth for calculating the zero plane displacement height, FAO
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** default : 0.67
    //                          ** min : 0
    //                          ** max : 1
    //                          ** unit : dimensionless
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547rl
    //            * name: plantHeight
    //                          ** description : plant Height
    //                          ** datatype : DOUBLE
    //                          ** default : 0
    //                          ** min : 0
    //                          ** max : 1000
    //                          ** unit : mm
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //            * name: wind
    //                          ** description : wind
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** default : 124000
    //                          ** min : 0
    //                          ** max : 1000000
    //                          ** unit : m/d
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
    //                          ** inputtype : variable
                //- outputs:
    //            * name: conductance
    //                          ** description : the boundary layer conductance
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** unit : m/d
    //                          ** uri : http://www1.clermont.inra.fr/siriusquality/?page_id=547
            double plantHeight = a.plantHeight;
            double wind = a.wind;
            double conductance;
            double h;
            h = Math.Max(10.0d, plantHeight) / 100.0d;
            conductance = wind * Math.Pow(vonKarman, 2) / (Math.Log((heightWeatherMeasurements - (d * h)) / (zm * h)) * Math.Log((heightWeatherMeasurements - (d * h)) / (zh * h)));
            s.conductance= conductance;
        }

    }
}